### PR TITLE
Avoid Invalid code suggested when encountering unsatisfied trait bounds in derive macro code

### LINKED ISSUE #104884: Invalid code suggested when encountering unsatisfied trait bounds in derive macro code
Given the following code: <https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0345475f1b234a0cf26f97bfb712cceb>

```rust
use std::collections::BinaryHeap;

#[derive(PartialEq, Eq, PartialOrd, Ord, serde::Serialize)]
struct PriorityQueueEntry<T> {
    value: T,
}

#[derive(serde::Serialize)]
struct PriorityQueue<T>(BinaryHeap<PriorityQueueEntry<T>>);
```

The current output is:

```
Compiling playground v0.0.1 (/playground)
error[[E0277]](https://doc.rust-lang.org/nightly/error-index.html#E0277): the trait bound `T: Ord` is not satisfied
   --> src/lib.rs:8:10
    |
8   | #[derive(serde::Serialize)]
    |          ^^^^^^^^^^^^^^^^ the trait `Ord` is not implemented for `T`
9   | struct PriorityQueue<T>(BinaryHeap<PriorityQueueEntry<T>>);
    |                         --------------------------------- required by a bound introduced by this call
    |
note: required for `PriorityQueueEntry<T>` to implement `Ord`
   --> src/lib.rs:3:37
    |
3   | #[derive(PartialEq, Eq, PartialOrd, Ord, serde::Serialize)]
    |                                     ^^^
    = note: required for `BinaryHeap<PriorityQueueEntry<T>>` to implement `Serialize`
note: required by a bound in `serialize_newtype_struct`
   --> /playground/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.147/src/ser/mod.rs:904:12
    |
904 |         T: Serialize;
    |            ^^^^^^^^^ required by this bound in `serialize_newtype_struct`
    = note: this error originates in the derive macro `Ord` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting type parameter `T`
    |
8   | #[derive(serde::Serialize, T: std::cmp::Ord)]
    |                          ++++++++++++++++++
```

The last part of the suggestion `#[derive(serde::Serialize, T: std::cmp::Ord)]` is not valid Rust code and therefore should not be suggested like this.

This problem occurs on the Playground on both stable and nightly with basically identical error messages. serde v1.0.147 requires `T: Ord` such that `BinaryHeap` will be `Serialize`. Otherwise the problem does not seem serde specific.

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #104884

### CODE REVIEW

**Comment by chenyukang** [comment]
I verified the change fixed this issue with `serde`, but I'm struggling to construct a minimal case without `serde`.

Appreciate if anyone could help me on this.

**Comment by petrochenkov** [comment]
r? rust-lang/diagnostics
Skipping suggestions for macro code is relevant for pretty much all reported diagnostics.
There's probably some centralized way to do this.

**Comment by petrochenkov** [comment]
>I'm struggling to construct a minimal case without `serde`.

See `src/test/ui/proc-macro` for examples of using proc macros in the test suite.

**Comment by TaKO8Ki** [comment]
Given the following code as a MVCE:

src/test/ui/proc-macro/add-trait-impl.rs
```rust
// aux-build:add-trait-impl.rs

use std::collections::BinaryHeap;

#[macro_use]
extern crate add_trait_impl;

#[derive(PartialEq, Eq, PartialOrd, Ord)]
struct PriorityQueueEntry<T> {
    value: T,
}

#[derive(PartialOrd, AddImpl)]
struct PriorityQueue<T>(BinaryHeap<PriorityQueueEntry<T>>);

fn main() {}
```

src/test/ui/proc-macro/auxiliary/add-trait-impl.rs
```rust
// force-host
// no-prefer-dynamic

#![crate_type = "proc-macro"]

extern crate proc_macro;

use proc_macro::TokenStream;

#[proc_macro_derive(AddImpl)]

pub fn derive(input: TokenStream) -> TokenStream {
    "use std::cmp::Ordering;

    impl<T> Ord for PriorityQueue<T> {
        fn cmp(&self, other: &Self) -> Ordering {
            self.0.cmp(&self.height)
        }
    }
    "
    .parse()
    .unwrap()
}
```

Current output is:

```
error[E0277]: can't compare `PriorityQueue<T>` with `PriorityQueue<T>`
  --> /Users/maeda.takayuki/GitHub/rust/rust/src/test/ui/proc-macro/add-trait-impl.rs:13:10
   |
LL | #[derive(PartialOrd, AddImpl)]
   |          ^^^^^^^^^^ no implementation for `PriorityQueue<T> == PriorityQueue<T>`
   |
   = help: the trait `PartialEq` is not implemented for `PriorityQueue<T>`
note: required by a bound in `PartialOrd`
  --> /Users/maeda.takayuki/GitHub/rust/rust/library/core/src/cmp.rs:1083:43
   |
LL | pub trait PartialOrd<Rhs: ?Sized = Self>: PartialEq<Rhs> {
   |                                           ^^^^^^^^^^^^^^ required by this bound in `PartialOrd`
   = note: this error originates in the derive macro `PartialOrd` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `PriorityQueue<T>: Eq` is not satisfied
  --> /Users/maeda.takayuki/GitHub/rust/rust/src/test/ui/proc-macro/add-trait-impl.rs:13:22
   |
LL | #[derive(PartialOrd, AddImpl)]
   |                      ^^^^^^^ the trait `Eq` is not implemented for `PriorityQueue<T>`
   |
note: required by a bound in `Ord`
  --> /Users/maeda.takayuki/GitHub/rust/rust/library/core/src/cmp.rs:765:16
   |
LL | pub trait Ord: Eq + PartialOrd<Self> {
   |                ^^ required by this bound in `Ord`
   = note: this error originates in the derive macro `AddImpl` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: can't compare `T` with `T`
  --> /Users/maeda.takayuki/GitHub/rust/rust/src/test/ui/proc-macro/add-trait-impl.rs:13:22
   |
LL | #[derive(PartialOrd, AddImpl)]
   |                      ^^^^^^^ no implementation for `T < T` and `T > T`
   |
note: required for `PriorityQueue<T>` to implement `PartialOrd`
  --> /Users/maeda.takayuki/GitHub/rust/rust/src/test/ui/proc-macro/add-trait-impl.rs:13:10
   |
LL | #[derive(PartialOrd, AddImpl)]
   |          ^^^^^^^^^^
note: required by a bound in `Ord`
  --> /Users/maeda.takayuki/GitHub/rust/rust/library/core/src/cmp.rs:765:21
   |
LL | pub trait Ord: Eq + PartialOrd<Self> {
   |                     ^^^^^^^^^^^^^^^^ required by this bound in `Ord`
   = note: this error originates in the derive macro `AddImpl` which comes from the expansion of the derive macro `PartialOrd` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting type parameter `T`
   |
LL | #[derive(PartialOrd, AddImpl: std::cmp::PartialOrd)]
   |                             ++++++++++++++++++++++

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0277`.
```

The suggestion is a little bit different from the issue says, but it might help you.

**Comment by chenyukang** [comment]
Thanks very much! @TaKO8Ki 
This code could be MCVE for the issue, seems `serde` does something different,
but the wrong suggestion comes with the same codepath, and this PR fixes it.
I will add it as unit testcase.

**Comment by TaKO8Ki** [comment]
After reviewing your code, I thought that rustc should suggest like the following in this case. What do you think about? That is because skipping suggestions for macro code is just a quick fix.

```rust
use std::collections::BinaryHeap;

#[derive(PartialEq, Eq, PartialOrd, Ord, serde::Serialize)]
struct PriorityQueueEntry<T> {
    value: T,
}

#[derive(serde::Serialize)]
struct PriorityQueue<T>(BinaryHeap<PriorityQueueEntry<T>>);
```

```diff
error[E0277]: the trait bound `T: Ord` is not satisfied
   --> src/main.rs:8:10
    |
8   | #[derive(serde::Serialize)]
    |          ^^^^^^^^^^^^^^^^ the trait `Ord` is not implemented for `T`
9   | struct PriorityQueue<T>(BinaryHeap<PriorityQueueEntry<T>>);
    |                         --------------------------------- required by a bound introduced by this call
    |
note: required for `PriorityQueueEntry<T>` to implement `Ord`
   --> src/main.rs:3:37
    |
3   | #[derive(PartialEq, Eq, PartialOrd, Ord, serde::Serialize)]
    |                                     ^^^
    = note: required for `BinaryHeap<PriorityQueueEntry<T>>` to implement `Serialize`
note: required by a bound in `serialize_newtype_struct`
   --> /Users/maeda.takayuki/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.148/src/ser/mod.rs:904:12
    |
904 |         T: Serialize;
    |            ^^^^^^^^^ required by this bound in `serialize_newtype_struct`
    = note: this error originates in the derive macro `Ord` (in Nightly builds, run with -Z macro-backtrace for more info)
- help: consider further restricting type parameter `T`
-     |
- 8   | #[derive(serde::Serialize, T: std::cmp::Ord)]
-     |                          ++++++++++++++++++
+ help: consider restricting type parameter `T`
+   |
+ 9 | struct PriorityQueue<T: std::cmp::Ord>(BinaryHeap<PriorityQueueEntry<T>>);
+   |                       +++++++++++++++
```

**Comment by chenyukang** [comment]
Yes, you're right!
seems the span for suggestion is wrong.
Let me think how to fix it.

**Comment by chenyukang** [comment]
@TaKO8Ki I'm still learning this part of code, 
if you already have a fix please create a new PR, I will close this one😊

**Comment by chenyukang** [comment]
I guess `bounds_span_for_suggestions` does not handle well for `derive` scenario.

File: compiler/rustc_middle/src/ty/diagnostics.rs [hunk 1]
```diff
@@ -355,6 +355,10 @@ pub fn suggest_constraining_type_params<'a>(
         ));
     }
 
+    // remove the suggestions that are not suitable for diagnostics
+    suggestions =
+        suggestions.into_iter().filter(|(span, _, _)| !span.from_expansion()).collect::<Vec<_>>();
```
Comment (TaKO8Ki) [inline compiler/rustc_middle/src/ty/diagnostics.rs:360]: Then, could you use `in_derive_expansion` instead and leave a FIXME comment?

```suggestion
    // remove the suggestions that are not suitable for diagnostics
    suggestions =
        suggestions.into_iter().filter(|(span, _, _)| !span.in_derive_expansion()).collect::<Vec<_>>();
```

`from_expansion` skips suggestions for all kinds of macro including bang macros.

**Pull request merged**